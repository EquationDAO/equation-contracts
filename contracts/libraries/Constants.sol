// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.0;

library Constants {
    uint32 internal constant BASIS_POINTS_DIVISOR = 100_000_000;

    uint16 internal constant ADJUST_FUNDING_RATE_INTERVAL = 1 hours;
    uint16 internal constant SAMPLE_PREMIUM_RATE_INTERVAL = 5 seconds;
    uint16 internal constant REQUIRED_SAMPLE_COUNT = ADJUST_FUNDING_RATE_INTERVAL / SAMPLE_PREMIUM_RATE_INTERVAL;
    /// @dev 8 * (1+2+3+...+720) = 8 * ((1+720) * 720 / 2) = 8 * 259560
    uint32 internal constant PREMIUM_RATE_AVG_DENOMINATOR = 8 * 259560;
    /// @dev RoundingUp(50000 / 8 * Q96 / BASIS_POINTS_DIVISOR) = 4951760157141521099596497
    int256 internal constant PREMIUM_RATE_CLAMP_BOUNDARY_X96 = 4951760157141521099596497; // 0.05% / 8

    uint8 internal constant VERTEX_NUM = 7;
    uint8 internal constant LATEST_VERTEX = VERTEX_NUM - 1;

    uint64 internal constant RISK_BUFFER_FUND_LOCK_PERIOD = 90 days;

    uint256 internal constant Q64 = 1 << 64;
    uint256 internal constant Q96 = 1 << 96;

    bytes32 internal constant ROLE_POSITION_LIQUIDATOR = keccak256("ROLE_POSITION_LIQUIDATOR");
    bytes32 internal constant ROLE_LIQUIDITY_POSITION_LIQUIDATOR = keccak256("ROLE_LIQUIDITY_POSITION_LIQUIDATOR");
}
