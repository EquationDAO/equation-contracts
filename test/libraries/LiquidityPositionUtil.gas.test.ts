import {ethers} from "hardhat";
import {loadFixture} from "@nomicfoundation/hardhat-network-helpers";
import {expectSnapshotGasCost} from "../shared/snapshotGasCost";
import {DECIMALS_18, DECIMALS_6, Q96, SIDE_LONG, SIDE_SHORT, toPriceX96} from "../shared/Constants";
import {IPoolLiquidityPosition} from "../../typechain-types/contracts/test/LiquidityPositionUtilTest";
import LiquidityPositionStruct = IPoolLiquidityPosition.LiquidityPositionStruct;
import GlobalUnrealizedLossMetricsStruct = IPoolLiquidityPosition.GlobalUnrealizedLossMetricsStruct;

describe("LiquidityPositionUtil gas tests", () => {
    async function deployFixture() {
        const LiquidityPositionUtil = await ethers.getContractFactory("LiquidityPositionUtil");
        const _liquidityPositionUtil = await LiquidityPositionUtil.deploy();
        await _liquidityPositionUtil.deployed();

        const LiquidityPositionUtilTest = await ethers.getContractFactory("LiquidityPositionUtilTest", {
            libraries: {
                LiquidityPositionUtil: _liquidityPositionUtil.address,
            },
        });
        const liquidityPositionUtil = await LiquidityPositionUtilTest.deploy();
        await liquidityPositionUtil.deployed();
        return {
            liquidityPositionUtil,
        };
    }

    describe("#updateUnrealizedLossMetrics", () => {
        describe("current unrealized loss is zero", () => {
            it("metrics before is zero", async () => {
                const {liquidityPositionUtil} = await loadFixture(deployFixture);
                await expectSnapshotGasCost(
                    liquidityPositionUtil.updateUnrealizedLossMetrics(
                        0n,
                        Math.floor(new Date().getTime() / 1000),
                        10n ** 18n,
                        Math.floor(new Date().getTime() / 1000),
                        0n
                    )
                );
            });

            it("metrics before is not zero", async () => {
                const {liquidityPositionUtil} = await loadFixture(deployFixture);
                await liquidityPositionUtil.updateUnrealizedLossMetrics(
                    123456n,
                    Math.floor(new Date().getTime() / 1000),
                    10n ** 18n,
                    Math.floor(new Date().getTime() / 1000),
                    123456n
                );

                await expectSnapshotGasCost(
                    liquidityPositionUtil.updateUnrealizedLossMetrics(
                        0n,
                        Math.floor(new Date().getTime() / 1000),
                        10n ** 18n,
                        Math.floor(new Date().getTime() / 1000),
                        0n
                    )
                );
            });
        });

        describe("current unrealized loss is not zero", () => {
            it("liquidity delta is zero", async () => {
                const {liquidityPositionUtil} = await loadFixture(deployFixture);
                await expectSnapshotGasCost(
                    liquidityPositionUtil.updateUnrealizedLossMetrics(
                        123456n,
                        Math.floor(new Date().getTime() / 1000),
                        0n,
                        Math.floor(new Date().getTime() / 1000),
                        123456n
                    )
                );
            });

            it("liquidity delta is positive", async () => {
                const {liquidityPositionUtil} = await loadFixture(deployFixture);
                await expectSnapshotGasCost(
                    liquidityPositionUtil.updateUnrealizedLossMetrics(
                        123456n,
                        Math.floor(new Date().getTime() / 1000),
                        10n ** 18n,
                        Math.floor(new Date().getTime() / 1000),
                        123456n
                    )
                );
            });

            it("liquidity delta is negative", async () => {
                const {liquidityPositionUtil} = await loadFixture(deployFixture);
                await liquidityPositionUtil.updateUnrealizedLossMetrics(
                    123456n,
                    Math.floor(new Date().getTime() / 1000),
                    10n ** 18n,
                    Math.floor(new Date().getTime() / 1000),
                    123456n
                );

                await expectSnapshotGasCost(
                    liquidityPositionUtil.updateUnrealizedLossMetrics(
                        123456n,
                        Math.floor(new Date().getTime() / 1000),
                        -(10n ** 18n),
                        Math.floor(new Date().getTime() / 1000),
                        123456n
                    )
                );
            });
        });
    });

    describe("#calculatePositionUnrealizedLoss", () => {
        it("entry time is greater than last zero loss time", async () => {
            const {liquidityPositionUtil} = await loadFixture(deployFixture);
            const lastTimestamp = Math.floor(new Date().getTime() / 1000);

            const globalPosition = newGlobalLiquidityPosition();
            globalPosition.liquidity = 1000n * 10n ** 6n;

            const position = newLiquidityPosition();
            position.liquidity = 333n * 10n ** 6n;
            position.entryTime = lastTimestamp;

            const metrics = newGlobalUnrealizedLossMetrics();
            metrics.lastZeroLossTime = lastTimestamp - 100;
            await expectSnapshotGasCost(
                liquidityPositionUtil.getGasCostCalculatePositionUnrealizedLoss(
                    position,
                    metrics,
                    globalPosition.liquidity,
                    100n * 10n ** 6n
                )
            );
        });

        describe("entry time is not greater than last zero loss time", () => {
            it("unrealized loss is greater than WAM unrealized loss", async () => {
                const {liquidityPositionUtil} = await loadFixture(deployFixture);
                const lastTimestamp = Math.floor(new Date().getTime() / 1000);

                const globalPosition = newGlobalLiquidityPosition();
                globalPosition.liquidity = 1000n * 10n ** 6n;

                const position = newLiquidityPosition();
                position.liquidity = 333n * 10n ** 6n;
                position.entryTime = lastTimestamp;

                const metrics = newGlobalUnrealizedLossMetrics();
                metrics.liquidity = 444n * 10n ** 6n;
                metrics.liquidityTimesUnrealizedLoss = metrics.liquidity * (20n * 10n ** 6n);
                metrics.lastZeroLossTime = lastTimestamp + 1;
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculatePositionUnrealizedLoss(
                        position,
                        metrics,
                        globalPosition.liquidity,
                        300n * 10n ** 6n
                    )
                );
            });

            it("unrealized loss is less than WAM unrealized loss", async () => {
                const {liquidityPositionUtil} = await loadFixture(deployFixture);
                const lastTimestamp = Math.floor(new Date().getTime() / 1000);

                const globalPosition = newGlobalLiquidityPosition();
                globalPosition.liquidity = 1000n * 10n ** 6n;

                const position = newLiquidityPosition();
                position.liquidity = 333n * 10n ** 6n;
                position.entryTime = lastTimestamp;

                const metrics = newGlobalUnrealizedLossMetrics();
                metrics.liquidity = 444n * 10n ** 6n;
                metrics.liquidityTimesUnrealizedLoss = metrics.liquidity * (200n * 10n ** 6n);
                metrics.lastZeroLossTime = lastTimestamp + 1;
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculatePositionUnrealizedLoss(
                        position,
                        metrics,
                        globalPosition.liquidity,
                        10n * 10n ** 6n
                    )
                );
            });
        });
    });

    describe("#calculateRealizedPnLAndNextEntryPriceX96", () => {
        it("the sum of netSize and liquidationBufferNetSize is zero", async () => {
            const {liquidityPositionUtil} = await loadFixture(deployFixture);

            const globalPosition = newGlobalLiquidityPosition();
            globalPosition.entryPriceX96 = toPriceX96("1808", DECIMALS_18, DECIMALS_6);
            {
                globalPosition.side = SIDE_SHORT;
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_SHORT,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        0n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_LONG,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        0n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_SHORT,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        1000n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_LONG,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        1000n
                    )
                );
            }

            {
                globalPosition.side = SIDE_LONG;
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_LONG,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        0n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_SHORT,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        0n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_LONG,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        1000n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_SHORT,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        1000n
                    )
                );
            }
        });

        it("the side of the trader's position adjustment is different from the side of global liquidity position", async () => {
            const {liquidityPositionUtil} = await loadFixture(deployFixture);

            const globalPosition = newGlobalLiquidityPosition();
            globalPosition.entryPriceX96 = toPriceX96("1808", DECIMALS_18, DECIMALS_6);
            globalPosition.netSize = Q96;
            globalPosition.liquidationBufferNetSize = Q96;

            {
                globalPosition.side = SIDE_SHORT;
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_SHORT,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        0n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_LONG,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        0n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_SHORT,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        1000n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_LONG,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        1000n
                    )
                );
            }

            {
                globalPosition.side = SIDE_LONG;
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_LONG,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        0n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_SHORT,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        0n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_LONG,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        1000n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_SHORT,
                        toPriceX96("1810", DECIMALS_18, DECIMALS_6),
                        1000n
                    )
                );
            }
        });

        it("the sum of netSize and liquidationBufferNetSize is positive and the side of the trader's position adjustment is same as the side of global liquidity position", async () => {
            const {liquidityPositionUtil} = await loadFixture(deployFixture);

            const globalPosition = newGlobalLiquidityPosition();
            globalPosition.entryPriceX96 = toPriceX96("1808", DECIMALS_18, DECIMALS_6);
            globalPosition.netSize = Q96;
            globalPosition.liquidationBufferNetSize = Q96;
            const tradePriceX96 = toPriceX96("1810", DECIMALS_18, DECIMALS_6);
            {
                globalPosition.side = SIDE_SHORT;
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_SHORT,
                        tradePriceX96,
                        Q96
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_SHORT,
                        tradePriceX96,
                        Q96 * 2n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_SHORT,
                        tradePriceX96,
                        Q96 * 3n
                    )
                );
            }

            {
                globalPosition.side = SIDE_LONG;
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_LONG,
                        tradePriceX96,
                        Q96
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_LONG,
                        tradePriceX96,
                        Q96 * 2n
                    )
                );
                await expectSnapshotGasCost(
                    liquidityPositionUtil.getGasCostCalculateRealizedPnLAndNextEntryPriceX96(
                        globalPosition,
                        SIDE_LONG,
                        tradePriceX96,
                        Q96 * 3n
                    )
                );
            }
        });
    });

    it("#increaseRiskBufferFundPosition", async () => {
        const {liquidityPositionUtil} = await loadFixture(deployFixture);

        const [owner, other] = await ethers.getSigners();
        await expectSnapshotGasCost(liquidityPositionUtil.getGasCostIncreaseRiskBufferFundPosition(owner.address, Q96));
        await expectSnapshotGasCost(liquidityPositionUtil.getGasCostIncreaseRiskBufferFundPosition(owner.address, 1n));
        await expectSnapshotGasCost(liquidityPositionUtil.getGasCostIncreaseRiskBufferFundPosition(other.address, 1n));
    });

    it("#decreaseRiskBufferFundPosition", async () => {
        const {liquidityPositionUtil} = await loadFixture(deployFixture);

        const [owner, other] = await ethers.getSigners();
        await liquidityPositionUtil.increaseRiskBufferFundPosition(owner.address, 1000n);
        await liquidityPositionUtil.increaseRiskBufferFundPosition(other.address, 100n);
        await expectSnapshotGasCost(
            liquidityPositionUtil.getGasCostDecreaseRiskBufferFundPosition(
                toPriceX96("1808.123", DECIMALS_18, DECIMALS_6),
                owner.address,
                1n
            )
        );
    });
});

function newGlobalLiquidityPosition() {
    return {
        netSize: 0n,
        liquidationBufferNetSize: 0n,
        entryPriceX96: 0n,
        side: 0,
        liquidity: 0n,
        realizedProfitGrowthX64: 0n,
    };
}

function newLiquidityPosition(): LiquidityPositionStruct {
    return {
        margin: 0n,
        liquidity: 0n,
        entryUnrealizedLoss: 0n,
        entryRealizedProfitGrowthX64: 0n,
        entryTime: 0,
        account: ethers.constants.AddressZero,
    };
}

function newGlobalUnrealizedLossMetrics(): GlobalUnrealizedLossMetricsStruct {
    return {
        lastZeroLossTime: 0n,
        liquidity: 0n,
        liquidityTimesUnrealizedLoss: 0n,
    };
}
