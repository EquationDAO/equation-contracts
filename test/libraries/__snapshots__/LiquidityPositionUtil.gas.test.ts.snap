// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiquidityPositionUtil gas tests #calculatePositionUnrealizedLoss entry time is greater than last zero loss time 1`] = `545`;

exports[`LiquidityPositionUtil gas tests #calculatePositionUnrealizedLoss entry time is not greater than last zero loss time unrealized loss is greater than WAM unrealized loss 1`] = `1154`;

exports[`LiquidityPositionUtil gas tests #calculatePositionUnrealizedLoss entry time is not greater than last zero loss time unrealized loss is less than WAM unrealized loss 1`] = `788`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the side of the trader's position adjustment is different from the side of global liquidity position 1`] = `751`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the side of the trader's position adjustment is different from the side of global liquidity position 2`] = `272`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the side of the trader's position adjustment is different from the side of global liquidity position 3`] = `783`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the side of the trader's position adjustment is different from the side of global liquidity position 4`] = `916`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the side of the trader's position adjustment is different from the side of global liquidity position 5`] = `673`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the side of the trader's position adjustment is different from the side of global liquidity position 6`] = `272`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the side of the trader's position adjustment is different from the side of global liquidity position 7`] = `673`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the side of the trader's position adjustment is different from the side of global liquidity position 8`] = `985`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is positive and the side of the trader's position adjustment is same as the side of global liquidity position 1`] = `751`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is positive and the side of the trader's position adjustment is same as the side of global liquidity position 2`] = `774`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is positive and the side of the trader's position adjustment is same as the side of global liquidity position 3`] = `1029`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is positive and the side of the trader's position adjustment is same as the side of global liquidity position 4`] = `673`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is positive and the side of the trader's position adjustment is same as the side of global liquidity position 5`] = `696`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is positive and the side of the trader's position adjustment is same as the side of global liquidity position 6`] = `950`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is zero 1`] = `226`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is zero 2`] = `226`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is zero 3`] = `482`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is zero 4`] = `481`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is zero 5`] = `226`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is zero 6`] = `226`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is zero 7`] = `481`;

exports[`LiquidityPositionUtil gas tests #calculateRealizedPnLAndNextEntryPriceX96 the sum of netSize and liquidationBufferNetSize is zero 8`] = `482`;

exports[`LiquidityPositionUtil gas tests #decreaseRiskBufferFundPosition 1`] = `46104`;

exports[`LiquidityPositionUtil gas tests #increaseRiskBufferFundPosition 1`] = `91921`;

exports[`LiquidityPositionUtil gas tests #increaseRiskBufferFundPosition 2`] = `40621`;

exports[`LiquidityPositionUtil gas tests #increaseRiskBufferFundPosition 3`] = `57721`;

exports[`LiquidityPositionUtil gas tests #updateUnrealizedLossMetrics current unrealized loss is not zero liquidity delta is negative 1`] = `26468`;

exports[`LiquidityPositionUtil gas tests #updateUnrealizedLossMetrics current unrealized loss is not zero liquidity delta is positive 1`] = `67229`;

exports[`LiquidityPositionUtil gas tests #updateUnrealizedLossMetrics current unrealized loss is not zero liquidity delta is zero 1`] = `24542`;

exports[`LiquidityPositionUtil gas tests #updateUnrealizedLossMetrics current unrealized loss is zero metrics before is not zero 1`] = `27569`;

exports[`LiquidityPositionUtil gas tests #updateUnrealizedLossMetrics current unrealized loss is zero metrics before is zero 1`] = `46669`;
